native/dlaqr5.go

```diff
@@ -0,0 +1,609 @@
+		if si[i] == -si[i+1] {
+			continue
+		}
+		swap := sr[i]
+		sr[i] = sr[i+1]
+		sr[i+1] = sr[i+2]
+		sr[i+2] = swap
+		swap = si[i]
+		si[i] = si[i+1]
+		si[i+1] = si[i+2]
+		si[i+2] = swap
+	}
+
+	// Note: lapack says that nshfts must be even but allows it to be odd
+	// anyway. At the moment, we panic above if nshfts is not even, so
+	// reducing it by one is unnecessary (it seems that Dlaqr0 and Dlaqr4
+	// indeed use only even nshfts).
+	//
+	// * NSHFTS is supposed to be even, but if it is odd,
+	// * then simply reduce it by one.  The shuffle above
+	// * ensures that the dropped shift is real and that
+	// * the remaining shifts are paired. ====
+	// NS = NSHFTS - MOD( NSHFTS, 2 )
+	ns := nshfts
+
+	safmin := dlamchS
+	ulp := dlamchP
+	smlnum := safmin * float64(n) / ulp
+
+	// Use accumulated reflections to update far-from-diagonal entries?
+	accum := kacc22 == 1 || kacc22 == 2
+	// If so, exploit the 2×2 block structure?
+	blk22 := ns > 2 && kacc22 == 2
+
+	// Clear trash.
+	if ktop+2 <= kbot {
+		h[(ktop+2)*ldh+ktop] = 0
+	}
+
+	// nbmps = number of 2-shift bulges in the chain.
+	nbmps := ns / 2
+
+	// kdu = width of slab.
+	kdu := 6*nbmps - 3
+
+	// Create and chase chains of nbmps bulges.
+	for incol := 3*(1-nbmps) + ktop - 1; incol <= kbot-2; incol += 3*nbmps - 2 {
+		ndcol := incol + kdu
+		if accum {
+			impl.Dlaset(blas.All, kdu, kdu, 0, 1, u, ldu)
+		}
+
+		// Near-the-diagonal bulge chase. The following loop performs
+		// the near-the-diagonal part of a small bulge multi-shift QR
+		// sweep. Each 6*nbmps-2 column diagonal chunk extends from
+		// column incol to column ndcol (including both column incol and
+		// column ndcol). The following loop chases a 3*nbmps column
+		// long chain of nbmps bulges 3*nbmps-2 columns to the right.
+		// (incol may be less than ktop and ndcol may be greater than
+		// kbot indicating phantom columns from which to chase bulges
+		// before they are actually introduced or to which to chase
+		// bulges beyond column kbot.)
+		for krcol := incol; krcol <= min(incol+3*nbmps-3, kbot-2); krcol++ {
+			// Bulges number mtop to mbot are active double implicit
+			// shift bulges. There may or may not also be small 2×2
+			// bulge, if there is room. The inactive bulges (if any)
+			// must wait until the active bulges have moved down the
+			// diagonal to make room. The phantom matrix paradigm
+			// described above helps keep track.
+
+			mtop := max(0, ((ktop-1)-krcol+2)/3)
+			mbot := min(nbmps, (kbot-krcol)/3) - 1
+			m22 := mbot + 1
+			bmp22 := (mbot < nbmps-1) && (krcol+3*m22 == kbot-2)
+
+			// Generate reflections to chase the chain right one
+			// column. (The minimum value of k is ktop-1.)
+			for m := mtop; m <= mbot; m++ {
+				k := krcol + 3*m
+				if k == ktop-1 {
+					impl.Dlaqr1(3, h[ktop*ldh+ktop:], ldh,
+						sr[2*m], si[2*m], sr[2*m+1], si[2*m+1],
+						v[m*ldv:m*ldv+3])
+					alpha := v[m*ldv]
+					_, v[m*ldv] = impl.Dlarfg(3, alpha, v[m*ldv+1:m*ldv+3], 1)
```

Comment:
Why is v sliced from 1 more here and in the next Dlarfg call than in the Dlaqr1 call above? They have the same index in the Fortran.